ttnn/cpp/ttnn/operations/data_movement.hpp

```diff
@@ -315,9 +319,115 @@ struct RepeatInterleave {
     }
 };
 
+struct Pad {
+    static inline const std::array<TensorSchema, 1> input_tensor_schemas() {
+        return {ttnn::TensorSchema{
+            2,  // min rank
+            4,  // max rank
+            {ttnn::bfloat16, ttnn::bfloat8_b, ttnn::uint16, ttnn::int32, ttnn::uint32},
+            {ttnn::TILE_LAYOUT},
+            true,   // can_be_on_device
+            false,  // can_be_on_cpu
+            false,  // can_be_scalar
+            false   // is_optional}
+        }};
+    }
+
+    template <typename... Args>
+    static auto input_tensors_to_validate(const ttnn::Tensor& input_tensor, Args&&... args) {
+        return std::make_tuple(input_tensor);
+    }
+
+    static ttnn::Tensor execute_on_worker_thread(
+        const ttnn::Tensor& input_tensor,
+        std::vector<std::pair<uint32_t, uint32_t>> padding, //intentionally not const&
+        const float value,
+        const std::optional<MemoryConfig>& memory_config_arg) {
+
+        const int original_rank = input_tensor.get_shape().rank();
+        if(int diff = original_rank - padding.size(); diff != 0) {
+            TT_FATAL(diff > 0, "ttnn.pad: padding len can't be larger than input tensor rank");
+
+            padding.insert(padding.begin(), diff, {0, 0});
+        }
+
+        TT_FATAL(
+            padding.size() == original_rank,
+            "ttnn.pad: padding must be the same length as the input tensor rank");
+        TT_FATAL(
+            input_tensor.get_layout() != ttnn::ROW_MAJOR_LAYOUT,
+            "ttnn.pad: row-major tensors have to use fallback because the kernel currently causes a PCC error");
+
+        // Unsqueeze Tensor to 4D if it is not already
+        ttnn::Tensor input_tensor_4D = input_tensor;
+        input_tensor_4D = ttnn::unsqueeze_to_4D(input_tensor);
```

Comment:
Seems like this makes redundant copy of the tensor during initialization. 

Substitute two lines with this one to avoid redundant copy: 
tnn::Tensor input_tensor_4d = ttnn::unsqueeze_to_4D(input_tensor); 

Example in sandbox (optimizer can't optimize redundant copy): https://godbolt.org/z/7ja8M9zjx

Reply:
Is it possible to modify shape of tensor without copying it?

Reply:
Good catch. Will do. Thanks!